# Insurance Risk Data Analysis & Preprocessing

### Insurance Dataset Variable Dictionary

<table>
<tr>
<td width="50%">


-  policy_id — Unique identifier for each insurance policy  
-  customer_age — Age of the policy holder  
-  region_code — Geographic region category  
-  region_density — Population density of region  
-  subscription_length — Policy duration  
-  segment — Customer/vehicle segment  
-  vehicle_age — Age of the vehicle  
-  model — Vehicle model type  
-  fuel_type — Fuel type (Petrol/Diesel/CNG)  
-  engine_type — Engine configuration  
-  displacement — Engine capacity (cc)  
-  cylinder — Number of cylinders  
-  max_torque — Engine torque output  
-  max_power — Engine horsepower  
-  transmission_type — Manual/Automatic  
-  steering_type — Steering system type  

</td>
<td width="50%">


-  length — Vehicle length  
-  width — Vehicle width  
-  gross_weight — Vehicle weight  
-  turning_radius — Turning capability  
-  airbags — Number of airbags  
-  is_esc — Electronic Stability Control  
-  is_tpms — Tire pressure monitoring system  
-  is_parking_sensors — Parking sensors  
-  is_parking_camera — Reverse camera  
-  is_brake_assist — Brake assist system  
-  is_speed_alert — Speed warning system  
-  is_power_steering — Power steering system  
-  is_central_locking — Central locking system  
-  is_ecw — Emergency collision warning  
-  is_driver_seat_height_adjustable — Seat adjustment  
-  rear_brakes_type — Brake system type  
-  ncap_rating — Safety rating (0–5)  
-  claim_status — Insurance claim outcome  

</td>
</tr>
</table>

### Importing the working libraries

In [1]:
import pandas as pd
import numpy as np

# File handling
from pathlib import Path

# Warnings control (keeps notebook clean)
import warnings
warnings.filterwarnings('ignore')

# Display settings (for better readability in notebooks)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.width', 1000)

print("✓ Core libraries loaded successfully")

✓ Core libraries loaded successfully


In [2]:
df= pd.read_csv("Insurance claims data.csv")

### Preliminary check block

In [3]:
print("COLUMN NAMES:")

# Print column names 10 per row for readability
columns = df.columns.tolist()
for i in range(0, len(columns), 10):
    print(columns[i:i+10])

print("=" * 100)

print("SHAPE (rows, columns):")
print(df.shape)

print("=" * 100)

print("DATA TYPES:")
print(df.dtypes)

print("=" * 100)

print("MISSING VALUES SUMMARY:")

missing_df = pd.DataFrame({
    'Missing_Count': df.isnull().sum(),
    'Missing_Percent': (df.isnull().sum() / len(df) * 100).round(2)
})

# Only show columns with missing values
missing_df = missing_df[missing_df['Missing_Count'] > 0] \
    .sort_values(by='Missing_Percent', ascending=False)

if missing_df.empty:
    print("No missing values found ✔")
else:
    display(missing_df)

print("=" * 100)

print("UNIQUE VALUES PER COLUMN:")
print(df.nunique())

print("=" * 100)

print("DESCRIPTIVE STATS (NUMERIC):")
display(df.describe())

print("=" * 100)

COLUMN NAMES:
['policy_id', 'subscription_length', 'vehicle_age', 'customer_age', 'region_code', 'region_density', 'segment', 'model', 'fuel_type', 'max_torque']
['max_power', 'engine_type', 'airbags', 'is_esc', 'is_adjustable_steering', 'is_tpms', 'is_parking_sensors', 'is_parking_camera', 'rear_brakes_type', 'displacement']
['cylinder', 'transmission_type', 'steering_type', 'turning_radius', 'length', 'width', 'gross_weight', 'is_front_fog_lights', 'is_rear_window_wiper', 'is_rear_window_washer']
['is_rear_window_defogger', 'is_brake_assist', 'is_power_door_locks', 'is_central_locking', 'is_power_steering', 'is_driver_seat_height_adjustable', 'is_day_night_rear_view_mirror', 'is_ecw', 'is_speed_alert', 'ncap_rating']
['claim_status']
SHAPE (rows, columns):
(58592, 41)
DATA TYPES:
policy_id                            object
subscription_length                 float64
vehicle_age                         float64
customer_age                          int64
region_code                    

,subscription_length,vehicle_age,customer_age,region_density,airbags,displacement,cylinder,turning_radius,length,width,gross_weight,ncap_rating,claim_status
count,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000,58592.000000
mean,6.111688,1.388473,44.823935,18826.858667,3.137066,1162.355851,3.626963,4.852893,3850.476891,1672.233667,1385.276813,1.759950,0.063968
std,4.142790,1.134413,6.935604,17660.174792,1.832641,266.304786,0.483616,0.228061,311.457119,112.089135,212.423085,1.389576,0.244698
min,0.000000,0.000000,35.000000,290.000000,1.000000,796.000000,3.000000,4.500000,3445.000000,1475.000000,1051.000000,0.000000,0.000000
25%,2.100000,0.400000,39.000000,6112.000000,2.000000,796.000000,3.000000,4.600000,3445.000000,1515.000000,1185.000000,0.000000,0.000000
50%,5.700000,1.200000,44.000000,8794.000000,2.000000,1197.000000,4.000000,4.800000,3845.000000,1735.000000,1335.000000,2.000000,0.000000
75%,10.400000,2.200000,49.000000,27003.000000,6.000000,1493.000000,4.000000,5.000000,3995.000000,1755.000000,1510.000000,3.000000,0.000000
max,14.000000,20.000000,75.000000,73430.000000,6.000000,1498.000000,4.000000,5.200000,4300.000000,1811.000000,1720.000000,5.000000,1.000000


The dataset has **58,592 rows and 41 columns**, with no missing values, which is a strong starting point. It contains a mix of numeric, categorical, and many binary Yes/No features related to vehicle safety and specifications. The target variable `claim_status` is binary and appears imbalanced, with most cases being non-claims. A few columns like `policy_id`, `max_torque`, and `max_power` will need cleaning or removal before modeling. Overall, the data is well-structured but requires preprocessing for ML readiness.


## Data Cleaning & Preprocessing

In this section, we transform the raw dataset into a clean, structured, and analysis-ready format. Although the dataset is largely complete with no missing values, several preprocessing steps are required to ensure consistency and suitability for machine learning tasks.

The cleaning process includes:
- Standardizing column names for uniformity and compatibility  
- Converting binary categorical variables (Yes/No) into numeric format (0/1)  
- Verifying and visualizing missing values across all features  
- Identifying categorical variables for future encoding  
- Ensuring proper data types across all columns  
- Generating summary statistics for continuous numeric variables  

These steps are essential to eliminate inconsistencies, improve data quality, and prepare the dataset for downstream analysis such as exploratory data analysis and predictive modeling.

In [5]:
# Standardize Column Names
# Standardize column names
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('-', '_')
)

print("✓ Cleaned column names:")
print(df.columns.tolist())

✓ Cleaned column names:
['policy_id', 'subscription_length', 'vehicle_age', 'customer_age', 'region_code', 'region_density', 'segment', 'model', 'fuel_type', 'max_torque', 'max_power', 'engine_type', 'airbags', 'is_esc', 'is_adjustable_steering', 'is_tpms', 'is_parking_sensors', 'is_parking_camera', 'rear_brakes_type', 'displacement', 'cylinder', 'transmission_type', 'steering_type', 'turning_radius', 'length', 'width', 'gross_weight', 'is_front_fog_lights', 'is_rear_window_wiper', 'is_rear_window_washer', 'is_rear_window_defogger', 'is_brake_assist', 'is_power_door_locks', 'is_central_locking', 'is_power_steering', 'is_driver_seat_height_adjustable', 'is_day_night_rear_view_mirror', 'is_ecw', 'is_speed_alert', 'ncap_rating', 'claim_status']


In [6]:
# Convert binary categorical columns

# Automatically detect binary Yes/No columns
binary_cols = [
    col for col in df.columns
    if df[col].dtype == 'object'
    and df[col].dropna().isin(['Yes', 'No']).all()
]

# Convert to 0/1
for col in binary_cols:
    df[col] = df[col].map({'Yes': 1, 'No': 0}).astype(int)

print(f"✓ Converted {len(binary_cols)} binary columns")

# Show sample
df[binary_cols[:5]].head()


✓ Converted 17 binary columns


,is_esc,is_adjustable_steering,is_tpms,is_parking_sensors,is_parking_camera
0,1,1,1,1,1
1,0,1,0,1,1
2,1,1,1,1,1
3,0,0,0,1,0
4,0,1,0,1,0


In [7]:
# Compute missing values
missing_info = df.isna().sum()
missing_percent = (missing_info / len(df)) * 100

# Build summary table
missing_df = pd.DataFrame({
    'feature': missing_info.index,
    'missing_count': missing_info.values,
    'missing_percentage': missing_percent.values
})

# Sort by missing percentage
missing_df = missing_df.sort_values(
    by='missing_percentage',
    ascending=False
).reset_index(drop=True)

# Split into two halves
mid = int(np.ceil(len(missing_df) / 2))
left = missing_df.iloc[:mid].reset_index(drop=True)
right = missing_df.iloc[mid:].reset_index(drop=True)

# Align both halves
if len(left) > len(right):
    for _ in range(len(left) - len(right)):
        right = pd.concat([
            right,
            pd.DataFrame([{
                'feature': '',
                'missing_count': '',
                'missing_percentage': ''
            }])
        ], ignore_index=True)

# Combine side-by-side
combined_df = pd.DataFrame({
    'feature_left': left['feature'],
    'missing_count_left': left['missing_count'],
    'missing_percentage_left': left['missing_percentage'],
    'feature_right': right['feature'],
    'missing_count_right': right['missing_count'],
    'missing_percentage_right': right['missing_percentage'],
})

# Styling function
def highlight_missing(val):
    if val == '' or pd.isna(val):
        return ''
    elif val > 50:
        return 'background-color: red; color: white; font-weight: bold'
    elif val > 20:
        return 'background-color: orange; font-weight: bold'
    elif val > 5:
        return 'background-color: yellow'
    return ''

# Apply styling safely
styled_df = combined_df.style.apply(
    lambda col: col.map(highlight_missing),
    subset=[
        'missing_percentage_left',
        'missing_percentage_right'
    ]
)

styled_df

,feature_left,missing_count_left,missing_percentage_left,feature_right,missing_count_right,missing_percentage_right
0,policy_id,0,0.000000,subscription_length,0,0.000000
1,transmission_type,0,0.000000,displacement,0,0.000000
2,turning_radius,0,0.000000,vehicle_age,0,0.000000
3,length,0,0.000000,customer_age,0,0.000000
4,width,0,0.000000,region_code,0,0.000000
5,gross_weight,0,0.000000,region_density,0,0.000000
6,is_front_fog_lights,0,0.000000,segment,0,0.000000
7,is_rear_window_wiper,0,0.000000,model,0,0.000000
8,is_rear_window_washer,0,0.000000,fuel_type,0,0.000000
9,is_rear_window_defogger,0,0.000000,max_torque,0,0.000000


- We had no missing in our data

In [8]:
# Identify remaining categorical columns
cat_cols = df.select_dtypes(include='object').columns.tolist()

print("Categorical columns to encode later:")
print(cat_cols)

Categorical columns to encode later:
['policy_id', 'region_code', 'segment', 'model', 'fuel_type', 'max_torque', 'max_power', 'engine_type', 'rear_brakes_type', 'transmission_type', 'steering_type']


In [9]:
# Numeric Summary Statistics
# Identify binary columns (0/1 after encoding)
binary_cols = [col for col in df.columns if df[col].nunique() == 2]

# Identify numeric columns
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()

# Exclude binary columns
continuous_numeric_cols = [col for col in numeric_cols if col not in binary_cols]

# Summary statistics
num_summary = df[continuous_numeric_cols].describe().T

num_summary

,count,mean,std,min,25%,50%,75%,max
subscription_length,58592.0,6.111688,4.142790,0.0,2.1,5.7,10.4,14.0
vehicle_age,58592.0,1.388473,1.134413,0.0,0.4,1.2,2.2,20.0
customer_age,58592.0,44.823935,6.935604,35.0,39.0,44.0,49.0,75.0
region_density,58592.0,18826.858667,17660.174792,290.0,6112.0,8794.0,27003.0,73430.0
airbags,58592.0,3.137066,1.832641,1.0,2.0,2.0,6.0,6.0
displacement,58592.0,1162.355851,266.304786,796.0,796.0,1197.0,1493.0,1498.0
turning_radius,58592.0,4.852893,0.228061,4.5,4.6,4.8,5.0,5.2
length,58592.0,3850.476891,311.457119,3445.0,3445.0,3845.0,3995.0,4300.0
width,58592.0,1672.233667,112.089135,1475.0,1515.0,1735.0,1755.0,1811.0
gross_weight,58592.0,1385.276813,212.423085,1051.0,1185.0,1335.0,1510.0,1720.0


The dataset shows a generally well-structured set of numeric features with no missing values. Most vehicle characteristics such as dimensions and engine displacement are tightly distributed, indicating standardized vehicle designs. However, vehicle age and region density show noticeable outliers and skewness, suggesting variability in risk exposure. Safety-related features like airbags and NCAP ratings are limited in range but likely important for predicting claims. Overall, the data is clean but contains a few skewed variables that may need transformation during modeling.


In [11]:
#### saving our data
df.to_csv("cleaned_insurance_data.csv", index=False)

print("✓ Cleaned data saved as 'cleaned_insurance_data.csv'")

✓ Cleaned data saved as 'cleaned_insurance_data.csv'
